### PR TITLE
Not a cache eviction: Document stores running state

### PR DESCRIPTION
**Pull request by ukstv**
- Cleans the Document from conflict resolution and doctype-specific properties
- Document wraps a running state
- `Document.loadAtCommit` becomes `document.rewind`

Extracted from the cache eviction branch. Have some value on its own.

I could not win over failing https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/core/src/__tests__/ceramic-anchor.test.ts#L262 with all the changes. This is a minimal starting point before the test starts failing.

### CODE REVIEW

File: packages/core/src/__tests__/ceramic-anchor.test.ts [hunk 1]
```diff
@@ -225,7 +226,7 @@ describe('Ceramic anchoring', () => {
     ])
     const controller = ceramic1.context.did.id
 
-    const doctype1 = await ceramic1.createDocument(DOCTYPE_TILE, { content: { x: 1 } }, { anchor: false, publish: false })
+    const doctype1 = await ceramic1.createDocument<TileDoctype>(DOCTYPE_TILE, { content: { x: 1 } }, { anchor: false, publish: false })
```
Comment (stbrody) [inline packages/core/src/__tests__/ceramic-anchor.test.ts:229]: Kind of annoying that you have to both specialize on the doctype subclass AND provide the doctype name string as an arg.  Maybe in the future we could remove the string doctype name argument and get it from the type specialization.  Mostly just an observation, no need to do anything on that here

**Review by stbrody — COMMENTED** [review]
This looks good overall, but I wasn't expecting so much work to be done on `Document` if the plan is to throw it out.  I guess I'm just wondering what the plan is to get from here to the desired end state from the spec.  Are you planning on turning `Document` **into** the `StateManager`?

**Comment by ukstv** [comment]
> This looks good overall, but I wasn't expecting so much work to be done on `Document` if the plan is to throw it out. I guess I'm just wondering what the plan is to get from here to the desired end state from the spec. Are you planning on turning `Document` **into** the `StateManager`?

This is the most reasonable way forward. If cleared from ancillary stuff, it _is_ a StateManager - holds a state, linearises operations on it. You have a bunch of tests for the whole logic. Would be unwise to throw this treasure trove away.

Two reasons for this particular pull request.
1. I basically feel ashamed of being stuck with one of the existing tests after having proper cache eviction. Decided to go forward with a part of the branch that does not cause trouble.
2. The changes in PR anyway were planned. Originally, for later, but all the ancillary code was so blinding and so coupled to Document creation logic, I could not move forward without moving it away.

File: packages/core/src/conflict-resolution.ts [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+  static fromState(dispatcher: Dispatcher, state: DocState): HistoryLog {
+    return new HistoryLog(
+      dispatcher,
+      state.log.map((_) => _.cid),
+    );
+  }
+
+  constructor(private readonly dispatcher: Dispatcher, readonly items: CID[]) {}
+
+  @Memoize()
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/conflict-resolution.ts]: makes sense.

File: packages/core/src/conflict-resolution.ts [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    const log = await fetchLog(this.dispatcher, tip, stateLog);
+    if (log.length) {
+      return this.applyLog(initialState, stateLog, log);
+    }
+  }
+
+  /**
+   * Return state at `commitId` version.
+   */
+  async rewind(initialState: DocState, commitId: CommitID) {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/conflict-resolution.ts]: done.

**Comment by stbrody** [comment]
> If cleared from ancillary stuff, it _is_ a StateManager

Okay cool, makes sense.  Just wanted to make sure I understood the plan for where this is headed.  Thanks for clarifying!

File: packages/core/src/document.ts [hunk 4]
```diff
@@ -741,17 +396,14 @@ class Document extends EventEmitter implements DocStateHolder {
...
+    await this._applyQueue.onIdle()
+    this.state$.complete();
   }
 
   /**
    * Serializes the document content
    */
   toString (): string {
-    return JSON.stringify(this._doctype.state.content)
+    return JSON.stringify(this.state$.value.content)
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/document.ts:407]: Ooh, I would like to not touch this at the moment. Have no idea what are the repercussions. Next PR, definitely.

**Comment by ukstv** [comment]
And, I consider this a case, where it makes sense to _merge_ instead of _squash_. A lot of changes, next person touches the code better be aware of the reasoning.

**Pull request merged**